src/transform/groovy/ga/ozli/minecraftmods/groovylicious/transform/registroid/RegistroidASTTransformer.groovy

```diff
@@ -0,0 +1,357 @@
+                drFields.push(dr)
+        }
+
+        List.copyOf(targetClass.properties).each { PropertyNode prop ->
+            addons.each {
+                if (it.supportedTypes.any { TransformUtils.isSubclass(prop.type, it) }) {
+                    it.process(annotationNode, targetClass, prop, this)
+                }
+            }
+        }
+
+        if (getMemberValue(annotationNode, 'includeInnerClasses')) {
+            targetClass.innerClasses.each {
+                registerInnerAddons(annotationNode, targetClass, it, addons)
+            }
+        }
+
+        drFields = drFields.stream().filter { it !== null }.toList()
+        if (hasDuplicates(drFields, targetClass)) return
+        drFields.each {
+            transformField(annotationNode, targetClass, it)
+        }
+    }
+
+    private void registerInnerAddons(final AnnotationNode annotationNode, final ClassNode targetClass, final InnerClassNode innerClass, final List<RegistroidAddon> addons) {
+        List.copyOf(innerClass.properties).each { PropertyNode prop ->
+            addons.each {
+                if (it.supportedTypes.any { TransformUtils.isSubclass(prop.type, it) }) {
+                    it.process(annotationNode, targetClass, prop, this)
+                }
+            }
+        }
+        innerClass.innerClasses.each { registerInnerAddons(annotationNode, targetClass, it, addons) }
+    }
+
+    private boolean hasDuplicates(List<FieldNode> fields, ClassNode targetClass) {
+        final Set<ClassNode> types = new HashSet<>()
+        for (final field : fields) {
+            // DeferredRegister<T>
+            final drType = field.type.genericsTypes[0].type
+            if (drType in types) {
+                addError("Registroid ambiguity: Found multiple DeferredRegisters pointing to the same type, ${drType.name}", targetClass)
+                return true
+            }
+            types.add(drType)
+        }
+        return false
+    }
+
+    private FieldNode generateDR(final ClassNode targetClass, final FieldNode registryField) {
+        if (registryField.type == RESOURCE_KEY_TYPE) {
+            makeResourceKey(targetClass, registryField)
+        } else if (registryField.type.isDerivedFrom(REGISTRY_TYPE)) {
+            makeRegistry(targetClass, registryField)
+        } else if (GeneralUtils.isOrImplements(registryField.type, FORGE_REGISTRY_TYPE)) {
+            makeForgeRegistry(targetClass, registryField)
+        } else {
+            addError("No known way of representing registry from ${registryField.owner.name}.${registryField.name}", registryField)
+            return null
+        }
+    }
+
+    private FieldNode makeResourceKey(final ClassNode targetClass, final FieldNode registryField) {
+        // ResourceKey<Registry<T>>
+        final type = registryField.type.genericsTypes[0].type.genericsTypes[0].type
+        final modId = ModRegistry.getData(targetClass.packageName)?.modId()
+        if (modId === null) {
```

Comment:
Probably simpler and faster to infer the modId and do this check once rather than spread across 3 methods that could be called multiple times.